### PR TITLE
SMARTS matching performance improvements

### PR DESCRIPTION
**Pull request by mattwthompson**
@lilyminium did the work (https://github.com/openforcefield/openff-toolkit/issues/971#issuecomment-859057461) - I'm opening it as a PR since I need this behavior added in order to trust my energy comparison tests.

Related #967 

- [ ] Toggle on slow tests with `True` arguments
- [ ] Ensure protein tests pass
- [ ] Add arguments to `.create_openmm_system` an `.create_force`
- [ ] Docstrings

### CODE REVIEW

**Comment by jeff231li** [comment]
Is there a way to control the new arguments in `find_matches` (`unique` and `match_heavy_first`) when we call the method `ForceField.create_openmm_system`?

**Comment by j-wags** [comment]
Hm, I think that both of these changes (`unique` and `match_heavy_first`), if they provide the same behavior, should default to `True` throughout the codebase. 

The two things I'd like to see to convince myself that they're behaving well are:
- [ ] The protein parameterization tests (both [di- and tri-peptides](https://github.com/openforcefield/amber-ff-porting) and the [T4 lysozyme example](https://github.com/openforcefield/openff-toolkit/pull/881)) running successfully with them defaulting to `True`
- [ ] The `--runslow` test suite passing with them defaulting to `True`

**Comment by lilyminium** [comment]
> Hm, I think that both of these changes (unique and match_heavy_first), if they provide the same behavior, should default to True throughout the codebase.

`unique=False, match_heavy_first=False` is status quo. Shifting `unique=True` would result in different answers for `find_matches` and the like -- e.g. in the example molecule, 1296 vs 1 match. https://github.com/openforcefield/openff-toolkit/issues/971#issuecomment-859057461

`unique=True` toggles enumeration of every combination of the same atoms. So a SMARTS pattern matching methane CH4 would only give (C, H1, H2, H3, H4) for `unique=True`. However, `unique=False` would return `[(C, H1, H2, H3, H4), (C, H1, H2, H4, H3), (C, H1, H3, H2, H4), (C, H1, H3, H4, H2), (C, H1, H4, H2, H3), (C, H1, H4, H3, H2), (C, H2, H1, H3, H4), (C, H2, H1, H4, H3), (C, H2, H3, H1, H4), (C, H2, H3, H4, H1), (C, H2, H4, H1, H3), (C, H2, H4, H3, H1), (C, H3, H1, H2, H4), (C, H3, H1, H4, H2), (C, H3, H2, H1, H4), (C, H3, H2, H4, H1), (C, H3, H4, H1, H2), (C, H3, H4, H2, H1), (C, H4, H1, H2, H3), (C, H4, H1, H3, H2), (C, H4, H2, H1, H3), (C, H4, H2, H3, H1), (C, H4, H3, H1, H2), (C, H4, H3, H2, H1)]`. 

I *think* `match_heavy_first=True|False` should give identical behaviour (currently RDKit only, I couldn't work out OpenEye). For each heavy match, it pins heavy atoms and enumerates though the remaining Hs.

**Comment by mattwthompson** [comment]
> Is there a way to control the new arguments in `find_matches` (`unique` and `match_heavy_first`) when we call the method `ForceField.create_openmm_system`?

@jeff231li No, but thanks for asking, since this question makes me realize how there would be some cost to wiring through another argument or two from `ForceField.create_openmm_system` to every one of the dozen or so parameter handlers.

**Comment by mattwthompson** [comment]
We met on this and decided:
* `unique=True` should be implemented and made the default for `LibraryChargeHandler`
* `unique=False` should remain for all other handlers
* `unique` is not to be exposed as an argument in `create_openmm_system` but will be added as an argument to `find_matches` and `chemical_environment_matches`
* `match_heavy_first=True` is probably fine, and has been stress tested, but _might_ cause issues and therefore we should split it out

**Comment by j-wags** [comment]
My profiling is showing about a 90% runtime reduction for whole-molecule librarycharge assignment using `uniquify=True`. Reproducing example here: https://github.com/openforcefield/openff-toolkit/issues/971#issuecomment-895602640

**Comment by mattwthompson** [comment]
Looks like `max_matches` is exposed inconsistently - where do we want this to exist as an argument, and where don't we?

**Comment by j-wags** [comment]
Limiting `max_matches` in this context is fairly dangerous, so I'd in favor of never exposing that as a parameter.

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 1]
```diff
@@ -2222,13 +2222,15 @@ def __init__(self, parameter_type, environment_match):
...
-    def find_matches(self, entity):
+    def find_matches(self, entity, unique=False):
         """Find the elements of the topology/molecule matched by a parameter type.
 
         Parameters
         ----------
         entity : openff.toolkit.topology.Topology
             Topology to search.
+        unique : bool, default=False
+            If True, SMARTS matching will enumerate every possible combination of atoms.
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:2233]:
```suggestion
            If False, SMARTS matching will enumerate every valid permutation of matching atoms. 
            If True, only one order of each unique match will be returned.
```

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 2]
```diff
@@ -2252,6 +2259,9 @@ def _find_matches(self, entity, transformed_dict_cls=ValenceDict):
             will determine how groups of atom indices are stored
             and accessed (e.g for angles indices should be 0-1-2
             and not 2-1-0).
+        unique : bool, default=False
+            If True, SMARTS matching will enumerate every possible combination of atoms.
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:2263]:
```suggestion
            If False, SMARTS matching will enumerate every valid permutation of matching atoms. 
            If True, only one order of each unique match will be returned.
```

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 3]
```diff
@@ -2241,7 +2243,12 @@ def find_matches(self, entity):
         #  Should we reduce its scope and have a check here to make sure entity is a Topology?
         return self._find_matches(entity)
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:2244]: (blocking) Do we need to thread the `unique` kwarg through here? Ditto for all other instances of this pattern in the other ParameterHandlers.

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 4]
```diff
@@ -3343,7 +3354,7 @@ def check_handler_compatibility(self, other_handler):
             tolerance_attrs=float_attrs_to_compare,
         )
 
-    def find_matches(self, entity):
+    def find_matches(self, entity, unique=False):
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:3357]: (blocking) (1) The `unique` kwags isn't used in this method. I think it needs to be added.

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 5]
```diff
@@ -4298,7 +4315,7 @@ def check_handler_compatibility(
             identical_attrs=string_attrs_to_compare + int_attrs_to_compare,
         )
 
-    def find_matches(self, entity):
+    def find_matches(self, entity, unique=False):
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:4318]: ditto (1)

File: openff/toolkit/utils/rdkit_wrapper.py [hunk 6]
```diff
@@ -1960,6 +1965,36 @@ def _find_smarts_matches(rdmol, smirks, aromaticity_model="OEAroModel_MDL"):
         """
         from rdkit import Chem
 
+        def _match_smarts_with_heavy_atoms_first(rdmol, qmol, match_kwargs):
```
Comment (j-wags) [inline openff/toolkit/utils/rdkit_wrapper.py:1968]:
```suggestion
        # This code is part of a possible performance optimization that hasn't been validated 
        # for production use yet.
        def _match_smarts_with_heavy_atoms_first(rdmol, qmol, match_kwargs):
```

**Review by j-wags — CHANGES_REQUESTED** [review]
Thanks -- This is looking really close to production-ready, though I've found a few places where I think a kwarg should be getting passed/possible typos. 

Also, could you add a pair of tests for `RDKTKW.find_smarts_matches` and `OETKW.find_smarts_matches` to ensure that matching `CCC` to propane returns two matches when `unique=False` and just one when `unique=True`?

File: openff/toolkit/typing/engines/smirnoff/parameters.py (hunk 3 above)
Comment (mattwthompson) [reply openff/toolkit/typing/engines/smirnoff/parameters.py:2245]: Yes, we do. I got wrapped up in out-of-scope details when thinking about how to wire it through. I think the argument is passed through each handler now; please verify that this is what you had in mind.

File: openff/toolkit/typing/engines/smirnoff/parameters.py [hunk 7]
```diff
@@ -4017,7 +4032,9 @@ def from_molecule(cls, molecule):
     _INFOTYPE = LibraryChargeType  # info type to store
     _DEPENDENCIES = [vdWHandler, ElectrostaticsHandler]
 
-    def find_matches(self, entity):
+    _KWARGS = ["unique"]
```
Comment (j-wags) [inline openff/toolkit/typing/engines/smirnoff/parameters.py:4035]: (not blocking) I don't think this should be here -- It's a list of strings that tells `ff.create_openmm_system(args, kwargs)` which kwargs are consumed by `LibraryChargeHandler.create_force` and so should be forwarded on. As is coded now there isn't special handling for the `unique` kwarg in `LibraryChargeHandler.create_force` so this isn't doing anything. I think it should be removed but may have overlooked a detail.

**Review by j-wags — APPROVED** [review]
This looks great. Thanks for iterating on it @mattwthompson, and thanks @lilyminium for pushing the idea to begin with! I suggested one more cleanup, other than that, please update the releasehistory (maybe file this under _New Features and Behavior Changed_ just in case there's unintended side effects?) and this should be good to merge!

File: openff/toolkit/typing/engines/smirnoff/parameters.py (hunk 7 above)
Comment (mattwthompson) [reply openff/toolkit/typing/engines/smirnoff/parameters.py:4035]: That sounds correct, thanks for catching this. Lost to history is a period of time in which I confused myself about which kwargs are passed to `find_matches` and `create_force`, and even a brief attempt at refactoring how each handler accepts kwargs.

**Pull request merged**